design/REGO_V2_PROPOSAL.md

```diff
@@ -39,82 +38,103 @@ proposed concepts thus far are:
 
 The proposed language has the following properties:
 
-* Policies are packaged into modules
-* At most one default package.
-* At most **one `main()` `@policy`** per module.
+* Rules are packaged into modules
+* The absence of qualified package name indicates the default package.
+* Rules are evaluated within the module until a decision is reached, unless
```

Comment:
Given multiple rules in the same package, which decision is picked? This problem shows up in the example below with the authenticated rule and the anonymous rule. A few options come to mind:

1) Main rule defines decision when package is queried.
2) Querying package returns a map of decisions (e.g., keyed by rule name.)

These aren't mutually exclusive. E.g., default behaviour could be (2) but if a main rule is defined, then that behaviour is overridden.

Alternatively, I could imagine using anonymous rules for this but I don't see a benefit over (1) and (2).

Reply:
At the time of this update, it seemed like the simplest evaluation models would be:

1. Evaluate all `@rule` statements until the first decision is reached.
2. Evaluate an `@rule` by name.

However, option (1) assumes uniform decision types and the definition of some external decision criteria, e.g. first true, first false, first allow, first deny, deny overrides. The `main()` evaluation model is more straightforward, but not necessarily the simplest user experience when it comes to defining simple rules. 

I agree this is an area for refinement.